CA2234: If Uri overload calls String overload, the check can suggest to modify the Uri overload to call itself

Using [Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers) v2.9.4 (Latest).

CA2234 triggers a warning in the following case, when I believe it should not:

- A class has a method that takes a String URI parameter, and an overload of that method that takes a Uri parameter.
- The latter method calls the former method, passing `uri.ToString()` to that method.

Example:

    public static string DataUriMediaType(string uri) {
       // ...
    }

    public static string DataUriMediaType(Uri uri) {
       return uri == null ? null : DataUriMediaType(uri.ToString());
    }

The warning has the following form:

    warning CA2234: Modify '*.DataUriMediaType(Uri)' to call '*.DataUriMediaType(Uri)' instead of '*.DataUriMediaType(string)'

--- a/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/PassSystemUriObjectsInsteadOfStrings.cs
+++ b/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/PassSystemUriObjectsInsteadOfStrings.cs
@@ -55,12 +55,7 @@ namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
                 }
 
                 var analyzer = new PerCompilationAnalyzer(c.Compilation, @string, uri, GetInvocationExpression);
-
-                // REVIEW: I need to do this thing because OperationAnalysisContext doesn't give me OwningSymbol
-                c.RegisterOperationBlockStartAction(sc =>
-                {
-                    sc.RegisterOperationAction(oc => analyzer.Analyze(oc, sc.OwningSymbol), OperationKind.Invocation);
-                });
+                c.RegisterOperationAction(analyzer.Analyze, OperationKind.Invocation);
             });
         }
 
@@ -86,13 +81,9 @@ namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
                 _expressionGetter = expressionGetter;
             }
 
-            public void Analyze(OperationAnalysisContext context, ISymbol owningSymbol)
+            public void Analyze(OperationAnalysisContext context)
             {
                 var invocation = (IInvocationOperation)context.Operation;
-                if (invocation.TargetMethod == null)
-                {
-                    return;
-                }
                 var method = invocation.TargetMethod;
 
                 // check basic stuff that FxCop checks. 
@@ -117,10 +108,6 @@ namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
                     return;
                 }
 
-                // REVIEW: why IOperation doesn't contain things like compilation and semantic model?
-                //         it seems wierd that I need to do this to get thsoe.
-                var model = _compilation.GetSemanticModel(context.Operation.Syntax.SyntaxTree);
-
                 var stringParameters = method.Parameters.GetParametersOfType(_string);
                 if (!stringParameters.Any())
                 {
@@ -136,7 +123,7 @@ namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
                 }
 
                 // now we make sure we actually have overloads that contains uri type parameter
-                var overloads = model.GetMemberGroup(node, context.CancellationToken).OfType<IMethodSymbol>();
+                var overloads = context.Operation.SemanticModel.GetMemberGroup(node, context.CancellationToken).OfType<IMethodSymbol>();
                 if (!overloads.HasOverloadWithParameterOfType(method, _uri, context.CancellationToken))
                 {
                     // no overload that contains uri as parameter
@@ -175,12 +162,13 @@ namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
                         }
 
                         // okay all other type match. check the main one
-                        if (overload.Parameters[index].Type?.Equals(_uri) == true)
+                        if (overload.Parameters[index].Type?.Equals(_uri) == true &&
+                            !Equals(overload, context.ContainingSymbol))
                         {
                             context.ReportDiagnostic(
                                 node.CreateDiagnostic(
                                     Rule,
-                                    owningSymbol.ToDisplayString(SymbolDisplayFormats.ShortSymbolDisplayFormat),
+                                    context.ContainingSymbol.ToDisplayString(SymbolDisplayFormats.ShortSymbolDisplayFormat),
                                     overload.ToDisplayString(SymbolDisplayFormats.ShortSymbolDisplayFormat),
                                     method.ToDisplayString(SymbolDisplayFormats.ShortSymbolDisplayFormat)));
 

--- a/src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/PassSystemUriObjectsInsteadOfStringsTests.cs
+++ b/src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/PassSystemUriObjectsInsteadOfStringsTests.cs
@@ -480,6 +480,22 @@ namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines.UnitTests
 ", GetCA2234BasicResultAt(6, 13, "A.Method()", "A.Method(Uri, Integer, String)", "A.Method(String, Integer, String)"));
         }
 
+        [Fact, WorkItem(2688, "https://github.com/dotnet/roslyn-analyzers/issues/2688")]
+        public void CA2234NoWarningInvocationInUriOverload()
+        {
+            VerifyCSharp(@"
+    using System;
+
+    public class A : IComparable
+    {
+        public static void Method(string uri) { }
+        public static void Method(Uri uri) { Method(uri.ToString()); }
+
+        public int CompareTo(object obj) { throw new NotImplementedException(); }
+    }
+");
+        }
+
         private static DiagnosticResult GetCA2234CSharpResultAt(int line, int column, params string[] args)
         {
             return GetCSharpResultAt(line, column, PassSystemUriObjectsInsteadOfStringsAnalyzer.Rule, args);

